pkg/providers/v1/aws.go

```diff
@@ -1543,6 +1543,11 @@ func extractIPv4NodeAddresses(instance *ec2.Instance) ([]v1.NodeAddress, error)
 
 	addresses := []v1.NodeAddress{}
 
+	// sort by device index so that the first address added to the addresses list is from the first (primary) device
+	sort.Slice(instance.NetworkInterfaces, func(i, j int) bool {
+		return *instance.NetworkInterfaces[i].Attachment.DeviceIndex < *instance.NetworkInterfaces[j].Attachment.DeviceIndex
```

Comment:
Can you add checks for a `nil` attachment and use `aws.IntValue` to dereference the index? The EC2 API should return sane values but we shouldn't crash if it doesn't

Reply:
Should I filter out interfaces where the attachment is `nil`, or should I just sort them to the end of the list?  I feel like filtering them out would be the right solution since they should have addresses associated if the interface isn't actually attached to an instance.

Reply:
The check here: https://github.com/kubernetes/cloud-provider-aws/pull/909/files/40202804f2189a7e3958261c8a1732bd8fd7686c#diff-97a31d0f193ea0b4e2c538595daf93405351054377458053a5010f6fdb97f4dcR1554-R1556

Should filter those out, so you just need to handle the potential `nil` pointer here. Start or end of the slice doesn't matter to me 😄

Reply:
I believe the updated code should just cause them to be sorted to the end of the list.  It avoids allocating another slice, and the code that pulls the addresses out should ignore them anyway.